src/Microsoft.Health.Fhir.Core/Features/Storage/FhirMemoryCache.cs

```diff
@@ -0,0 +1,186 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Runtime.Caching;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Health.Fhir.Core.Features.Storage
+{
+    public sealed class FhirMemoryCache<T> : IMemoryCache<T>
+    {
+        private const int DefaultLimitSizeInMegabytes = 50;
+        private const int DefaultExpirationTimeInMinutes = 24 * 60;
+
+        private readonly string _cacheName;
+        private readonly ILogger _logger;
+        private readonly ObjectCache _cache;
+        private readonly TimeSpan _expirationTime;
+        private readonly bool _ignoreCase;
+        private readonly object _lock;
+
+        public FhirMemoryCache(string name, ILogger logger, bool ignoreCase = false)
+            : this(
+                  name,
+                  limitSizeInMegabytes: DefaultLimitSizeInMegabytes,
+                  expirationTime: TimeSpan.FromMinutes(DefaultExpirationTimeInMinutes),
+                  logger)
+        {
+        }
+
+        public FhirMemoryCache(string name, int limitSizeInMegabytes, TimeSpan expirationTime, ILogger logger, bool ignoreCase = false)
+        {
+            EnsureArg.IsNotNull(name, nameof(name));
+            EnsureArg.IsGt(limitSizeInMegabytes, 0, nameof(name));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _cacheName = name;
+
+            _cache = new MemoryCache(
+                _cacheName,
+                new NameValueCollection()
+                {
+                    { "CacheMemoryLimitMegabytes", limitSizeInMegabytes.ToString() },
+                });
+
+            _expirationTime = expirationTime;
+
+            _logger = logger;
+
+            _ignoreCase = ignoreCase;
+
+            _lock = new object();
+        }
+
+        public long CacheMemoryLimit => ((MemoryCache)_cache).CacheMemoryLimit;
+
+        /// <summary>
+        /// Get or add the value to cache.
+        /// </summary>
+        /// <typeparam name="T">Type of the value in cache</typeparam>
+        /// <param name="key">Key</param>
+        /// <param name="value">Value</param>
+        /// <returns>Value in cache</returns>
+        public T GetOrAdd(string key, T value)
+        {
+            key = FormatKey(key);
+
+            lock (_lock)
+            {
+                if (_cache.Contains(key))
+                {
+                    return (T)_cache[key];
+                }
+
+                AddInternal(key, value);
+            }
```

Comment:
Is there a `_cache.AddOrGetExisting`?

Reply:
Yes, there is. And reading the docs, I believe this class can be simpler. PR is updated.